Replace the inhouse collection dependency resolver with `resolvelib`

webknjaz opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This change replaces our in-tree dependency resolution with external `resolvelib` that comes from the pip ecosystem. In doing so, it substantially refactors `ansible-galaxy collection [download|install|list|verify]` subcommands.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
galaxy cli

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

~Sanity checks blocker: https://github.com/ansible-community/antsibull/issues/217~

This PR does not fully address #71784.

webknjaz committed:
Add resolvelib to direct ansible-core dependencies

This is a dependency resolver that pip 20.3+ uses by default.
We integrate it for use for the collection dependency resolution
in ansible-galaxy CLI.

Refs:
* https://github.com/sarugaku/resolvelib
* https://pypi.org/project/resolvelib/
* https://pradyunsg.me/blog/2020/03/27/pip-resolver-testing/

Co-authored-by: Matt Clay <matt@mystile.com>

webknjaz committed:
Add a multi Galaxy API proxy

webknjaz committed:
Make ansible-galaxy collection CLI use resolvelib

This is a disturbingly huge change that includes:

  * Refactoring of the `ansible-galaxy collection` CLI.
    In particular, it:

    - reimplements the `download`, `install` and `verify` subcommands
      from scratch;

    - reuses helper bits previously moved out into external modules;

    - replaces the old in-house resolver with a more clear
      implementation based on the resolvelib library[1][2].

  * Aligning the tests with the new expectations;

  * Adding more tests, integration ones in particular.

[1]: https://github.com/sarugaku/resolvelib
[2]: https://pradyunsg.me/blog/2020/03/27/pip-resolver-testing/

Co-authored-by: Jordan Borean <jborean93@gmail.com>
Co-authored-by: Sam Doran <sdoran@redhat.com>
Co-authored-by: Sloane Hertel <shertel@redhat.com>
Co-authored-by: Sviatoslav Sydorenko <webknjaz@redhat.com>

Signed-Off-By: Sviatoslav Sydorenko <webknjaz@redhat.com>

webknjaz committed:
Add artifacts manager

webknjaz commented on lib/ansible/collections/__init__.py:
```diff
@@ -8,7 +8,7 @@
 from ansible.module_utils._text import to_bytes
 
-B_FLAG_FILES = frozenset([b'MANIFEST.json', b'galaxy.yml'])
+B_FLAG_FILES = frozenset([b'MANIFEST.json', b'galaxy.yml', b'galaxy.yaml'])  # FIXME: this is inconsistent with what ansible-galaxy CLI does
```
(unrelated)

jborean93 replied:
Will you be removing this code because it's unrelated. Sounds like it should be in a separate PR.

webknjaz replied:
exactly

webknjaz replied:
https://github.com/ansible/ansible/pull/72665

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,401 @@
+        return self.type == 'file' or (
+            not self.type and
+            _is_file_path(self.src)
+        )
+
+    @property
+    def is_dir(self):
+        return self.type == 'dir' or (
+            not self.type and
+            _is_collection_dir(self.src)
+        )
+
+    @property
+    def namespace_collection_paths(self):
+        assert self.is_subdirs, 'This property only makes sense for namespaces'
+        return [
+            to_native(path)
+            for path in _find_collections_in_subdirs(self.src)
+        ]
+
+    @property
+    def is_subdirs(self):
+        return self.type == 'subdirs' or (
+            not self.type and
+            _is_collection_namespace_dir(self.src)
+        )
+
+    @property
+    def is_url(self):
+        return self.type == 'url' or (
+            not self.type and
+            _is_http_url(self.src)
+        )
+
+    @property
+    def is_scm(self):
+        return self.type == 'git' or (
+            not self.type and
+            _is_git_url(self.src)
+        )
```
FIXME: The `or ...` part should probably be removed. There's no reason for this legacy approach to be kept + we only use custom constructors that always set the correct type anyway.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,401 @@
+from ansible.module_utils.six import raise_from
+
+
+try:  # NOTE: py3/py2 compat
+    # FIXME: put somewhere into compat
+    _is_py_id = str.isidentifier
+except AttributeError:  # Python 2
+    # FIXME: port this to AnsibleCollectionRef.is_valid_collection_name
+    from re import match as _match_pattern
+    from tokenize import Name as _VALID_IDENTIFIER_REGEX
+    _valid_identifier_string_regex = ''.join((_VALID_IDENTIFIER_REGEX, r'\Z'))
+    def _is_py_id(tested_str):
+        # Ref: https://stackoverflow.com/a/55802320/595220
+        return bool(_match_pattern(_valid_identifier_string_regex, tested_str))
```
@nitzmahone is this something `AnsibleCollectionRef.is_valid_collection_name` would benefit from?

webknjaz replied:
Ref https://github.com/ansible/ansible/pull/73279

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,401 @@
+
+
+def _is_file_path(tested_str):
+    return os.path.isfile(to_bytes(tested_str, errors='surrogate_or_strict'))
+
+
+def _is_http_url(tested_str):
+    return urlparse(tested_str).scheme.lower() in {'http', 'https'}
+
+
+def _is_git_url(tested_str):
+    if isinstance(tested_str, bytes):
+        breakpoint()
+    return tested_str.startswith(('git+', 'git@'))
+
+
+def _is_concrete_artifact_pointer(tested_str):
+    return any(
+        predicate(tested_str)
+        for predicate in (
+            # NOTE: Maintain the checks to be sorted from light to heavy:
+            _is_git_url,
+            _is_http_url,
+            _is_file_path,
+            _is_collection_dir,
+            _is_collection_namespace_dir,
+        )
+    )
+
+
+def _is_fqcn(tested_str):
+    # FIXME: port this to AnsibleCollectionRef.is_valid_collection_name
+    if tested_str.count('.') != 1:
+        return False
+
+    return all(
+        # FIXME: keywords and identifiers are different in differnt Pythons
+        not iskeyword(ns_or_name) and _is_py_id(ns_or_name)
+        for ns_or_name in tested_str.split('.')
+    )
```
FIXME: All of these are supposed to operate on immutable inputs and so we should probably apply some sort of caching (like LRU).

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,401 @@
+                ' collection directory.',
+            )
+
+        tmp_inst_req = cls(None, None, dir_path, 'dir')
+        req_name = art_mgr.get_direct_collection_fqcn(tmp_inst_req)
+        req_version = art_mgr.get_direct_collection_version(tmp_inst_req)
```
FIXME: think about handling this hack better

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -559,34 +218,45 @@ def build_collection(collection_path, output_path, force):
+        ver=collection_meta['version'],
+    )
+    b_collection_output = os.path.join(
+        output_path,
```
This needs to be a byte string so you aren't trying to join a str and bytes together.

```suggestion
        to_bytes(output_path, errors='surrogate_or_strict'),
```

s-hertel replied:
fixed by passing bytes to this function

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -681,29 +407,125 @@ def install_collections(collections, output_path, apis, validate_certs, ignore_e
-                    else:
-                        raise
+    if force_deps:
+        breakpoint()
```
```suggestion
```

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -681,29 +407,125 @@ def install_collections(collections, output_path, apis, validate_certs, ignore_e
+            allow_pre_release=allow_pre_release,
+        )
+
+    breakpoint()
```
```suggestion
```

jborean93 commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,401 @@
+
+def _is_git_url(tested_str):
+    if isinstance(tested_str, bytes):
+        breakpoint()
```
```suggestion
```

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -681,29 +407,125 @@ def install_collections(collections, output_path, apis, validate_certs, ignore_e
+    )
+
+    if not requested_requirements:
+        display.display(
```
I'm fine with the change in message, just noting that this test needs to be fixed up to check the new assertion https://github.com/ansible/ansible/blob/cf39d9de258cb9c47de9043e1a85e327e177dba7/test/integration/targets/ansible-galaxy-collection/tasks/install.yml#L43.

jborean93 commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,401 @@
+        return to_native(self.__unicode__())
+
+
+    def __unicode__(self):
```
Is this even needed, we will be dropping py2 and `to_text` does not call `__unicode__`. You may as well just put this in `__str__`.

webknjaz replied:
I just want this interface to be complete. Also, if an unwrapped instance slips into the interpolation, it'll probably get called.

jborean93 commented on test/lib/ansible_test/_data/requirements/units.txt:
```diff
@@ -5,3 +5,4 @@ pytest
 pytest-mock
 pytest-xdist
 pyyaml
+resolvelib
```
This also needs to be in the `integration.txt` file in this folder.

jborean93 reviewed (commented):
Have added a few suggestions to fix up some of the test failures when trying `ansible-test ansible-galaxy-collection --docker` straight from the PR.

There are 3 failures so far that we need to look into a bit further

* Explicit version in cmd that is a pre-release is not found
    * e.g. `ansible-galaxy collection install namespace.name:1.1.1-beta.1`
    * https://github.com/ansible/ansible/blob/cf39d9de258cb9c47de9043e1a85e327e177dba7/test/integration/targets/ansible-galaxy-collection/tasks/install.yml#L61
* Grandchildren deps are not being found/installed
    * https://github.com/ansible/ansible/blob/cf39d9de258cb9c47de9043e1a85e327e177dba7/test/integration/targets/ansible-galaxy-collection/tasks/install.yml#L96
    * In that test `child_dep.child_dep2` is not being installed (`parent_dep.parent_collection` -> `child_dep.child_collection` -> this collection)
* The error message from the current resolver tells you what collection failed to be resolved and the version requirements. It seems like the current code just shows the collection defined in the CLI
    * https://github.com/ansible/ansible/blob/cf39d9de258cb9c47de9043e1a85e327e177dba7/test/integration/targets/ansible-galaxy-collection/tasks/install.yml#L129

There are a few `breakpoint()` and `assert` entries in the code but I assume they are just there for testing and will be removed once the PR is ready.

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -732,80 +555,103 @@ def validate_collection_path(collection_path):
 
 @contextmanager
-def _display_progress():
+def _display_progress(msg=None):
```
This change is now in https://github.com/ansible/ansible/pull/72663

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -732,80 +555,103 @@ def validate_collection_path(collection_path):
 
 @contextmanager
-def _display_progress():
+def _display_progress(msg=None):
     config_display = C.GALAXY_DISPLAY_PROGRESS
     display_wheel = sys.stdout.isatty() if config_display is None else config_display
 
+    global display
+    if msg is not None:
+        display.display(msg)
+
```
This change is now in https://github.com/ansible/ansible/pull/72663

webknjaz commented:
One of the sanity checks fails because of https://github.com/ansible-community/antsibull/issues/214

s-hertel committed:
Drop `ansible-galaxy collection verify` unit tests

webknjaz commented:
> The test `ansible-test sanity --test docs-build` [[explain](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/docs-build.html)] failed with the error:
> 
> ```
> Command "/usr/bin/python3.6 /root/ansible/test/sanity/code-smell/docs-build.py" returned exit status 1.
> >>> Standard Error
> Command 'make base_singlehtmldocs' failed with status code: 2
> --> Standard Output
> ../../hacking/build-ansible.py collection-meta --template-file=../templates/collections_galaxy_meta.rst.j2 --output-dir=rst/dev_guide/ ../../lib/ansible/galaxy/data/collections_galaxy_meta.yml
> ../../hacking/build-ansible.py document-config --template-file=../templates/config.rst.j2 --output-dir=rst/reference_appendices/ ../../lib/ansible/config/base.yml
> mkdir -p rst/cli
> ../../hacking/build-ansible.py generate-man --template-file=../templates/cli_rst.j2 --output-dir=rst/cli/ --output-format rst ../../lib/ansible/cli/*.py
> ../../hacking/build-ansible.py document-keywords --template-dir=../templates --output-dir=rst/reference_appendices/ ../../lib/ansible/keyword_desc.yml
> ../../hacking/build-ansible.py docs-build base -o rst ;\
> 
> Makefile:129: recipe for target 'base_plugins' failed
> --> Standard Error
> Traceback (most recent call last):
>   File "../../hacking/build-ansible.py", line 103, in <module>
>     main()
>   File "../../hacking/build-ansible.py", line 92, in main
>     retval = command.main(args)
>   File "/root/ansible/hacking/build_library/build_ansible/command_plugins/docs_build.py", line 164, in main
>     return generate_base_docs(args)
>   File "/root/ansible/hacking/build_library/build_ansible/command_plugins/docs_build.py", line 57, in generate_base_docs
>     '--dest-dir', args.output_dir])
>   File "/usr/local/lib/python3.6/dist-packages/antsibull/cli/antsibull_docs.py", line 277, in run
>     return ARGS_MAP[args.command]()
>   File "/usr/local/lib/python3.6/dist-packages/antsibull/cli/doc_commands/stable.py", line 390, in generate_docs
>     generate_docs_for_all_collections(venv, collection_dir, app_ctx.extra['dest_dir'], flog)
>   File "/usr/local/lib/python3.6/dist-packages/antsibull/cli/doc_commands/stable.py", line 269, in generate_docs_for_all_collections
>     venv, collection_dir, collection_names=collection_names))
>   File "/usr/lib/python3.6/asyncio/base_events.py", line 484, in run_until_complete
>     return future.result()
>   File "/usr/local/lib/python3.6/dist-packages/antsibull/docs_parsing/parsing.py", line 45, in get_ansible_plugin_info
>     venv, collection_dir, collection_names=collection_names)
>   File "/usr/local/lib/python3.6/dist-packages/antsibull/docs_parsing/ansible_internal.py", line 56, in get_ansible_plugin_info
>     collection_enum_cmd = venv_python(*collection_enum_args, _env=env)
>   File "/usr/local/lib/python3.6/dist-packages/sh.py", line 1549, in __call__
>     return RunningCommand(cmd, call_args, stdin, stdout, stderr)
>   File "/usr/local/lib/python3.6/dist-packages/sh.py", line 793, in __init__
>     self.wait()
>   File "/usr/local/lib/python3.6/dist-packages/sh.py", line 849, in wait
>     self.handle_command_exit_code(exit_code)
>   File "/usr/local/lib/python3.6/dist-packages/sh.py", line 877, in handle_command_exit_code
>     raise exc
> sh.ErrorReturnCode_1: 
> 
>   RAN: /tmp/tmpv_nyurk9/ansible-base-venv/bin/python /tmp/tmph7reqgvt
> 
>   STDOUT:
> 
> 
>   STDERR:
> Traceback (most recent call last):
>   File "/tmp/tmph7reqgvt", line 19, in <module>
>     from ansible.galaxy.collection import CollectionRequirement
> ImportError: cannot import name 'CollectionRequirement'
> 
> make: *** [base_plugins] Error 1
> ```

This will be addressed by https://github.com/ansible-community/antsibull/pull/215.

webknjaz committed:
Drop-in dependency resolver implementation

Co-authored-by: Jordan Borean <jborean93@gmail.com>
Co-authored-by: Sloane Hertel <shertel@redhat.com>
Co-authored-by: Sviatoslav Sydorenko <webknjaz@redhat.com>

samdoran commented:
!waffling core_review

s-hertel commented on lib/ansible/cli/galaxy.py:
```diff
@@ -791,25 +794,31 @@ def execute_build(self):
             collection_path = GalaxyCLI._resolve_path(collection_path)
             build_collection(collection_path, output_path, force)
```
other way to fix it
```suggestion
            build_collection(collection_path, b_output_path, force)
```

webknjaz replied:
Yeah, that's what I was going to do.

samdoran commented:
!waffling needs_revision

webknjaz commented on test/integration/targets/ansible-galaxy-collection/files/test_module.py:
```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/python
```
sanity check doesn't like this ^

s-hertel replied:
```suggestion
```

webknjaz committed:
Add a dev representation method to GalaxyAPI

jborean93 committed:
Remove superfluous test for API servers fallbacks

s-hertel commented on lib/ansible/galaxy/collection/concrete_artifact_manager.py:
```diff
@@ -0,0 +1,602 @@
+            except LookupError as lookup_err:
+                raise_from(
+                    AnsibleError(
+                        'Failed to find the collection dir deps: {err!s}'.
```
If I have a collection in my install path that doesn't have a MANIFEST.json or galaxy.yml and try to install an unrelated collection, I get `ERROR! Failed to find the collection dir deps: The collection galaxy.yml path '/home/fedora/ansible/test_collections/ansible_collections/test/coll/galaxy.yml' does not exist.` We could probably just determine the namespace and name from the path, and the version would be "*" and no dependencies.

webknjaz replied:
Could you add an integration test for that in a PR so I could attempt to see what state it is in, in runtime?

s-hertel commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -684,27 +412,122 @@ def install_collections(collections, output_path, apis, validate_certs, ignore_e
+    # NOTE: Don't attempt to reevaluate already installed deps
+    # NOTE: unless `--force` is passed
+    # FIXME: debug why `--force` doesn't always work well
+    requested_requirements -= set() if force else {
```
`--force-with-deps` implies `--force` on devel. I'm not sure if we want to fix that by checking for either, e.g. 
```suggestion
    requested_requirements -= set() if (force or force_deps) else {
```
or if we should set `force` to True in cli/galaxy.py if `--force-with-deps` was provided (might be more minimal that way, since it looks like there are a few places we check for `force`).

webknjaz replied:
I'm about to rewrite a few bits around this logic because this var is no longer passed to the resolver and we still need some fixes related to SCM that would probably need to touch this now.

webknjaz committed:
Fix `is_pre_release` to treat `*` as non-pre

webknjaz committed:
Require `--force` for updating collections

s-hertel commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,457 @@
+
+        try:
+            return cls.from_dir_path(dir_path, art_mgr)
+        except ValueError:
```
`from_dir_path` tries to get the FQCN from the metadata and can cause an IOError if the metadata file doesn't exist:
```
Traceback (most recent call last):
  File "/root/ansible/bin/ansible-galaxy", line 135, in <module>
    exit_code = cli.run()
  File "/root/ansible/lib/ansible/cli/galaxy.py", line 536, in run
    context.CLIARGS['func']()
  File "/root/ansible/lib/ansible/cli/galaxy.py", line 1299, in execute_list
    self.execute_list_collection()
  File "/root/ansible/lib/ansible/cli/galaxy.py", line 68, in method_wrapper
    return wrapped_method(*args, **kwargs)
  File "/root/ansible/lib/ansible/cli/galaxy.py", line 1423, in execute_list_collection
    collection_path, artifacts_manager,
  File "/root/ansible/lib/ansible/galaxy/collection/__init__.py", line 993, in find_existing_collections
    artifacts_manager,
  File "/root/ansible/lib/ansible/galaxy/dependency_resolution/dataclasses.py", line 162, in from_dir_path_as_unknown
    return cls.from_dir_path(dir_path, art_mgr)
  File "/root/ansible/lib/ansible/galaxy/dependency_resolution/dataclasses.py", line 187, in from_dir_path
    req_name = art_mgr.get_direct_collection_fqcn(tmp_inst_req)
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 206, in get_direct_collection_fqcn
    self._get_direct_collection_namespace(collection),  # type: ignore
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 190, in _get_direct_collection_namespace
    return self.get_direct_collection_meta(collection)['namespace']  # type: ignore
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 234, in get_direct_collection_meta
    collection_meta = _get_meta_from_dir(b_artifact_path)
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 477, in _get_meta_from_dir
    return _get_meta_from_installed_dir(b_path)
  File "/root/ansible/lib/ansible/galaxy/collection/concrete_artifact_manager.py", line 519, in _get_meta_from_installed_dir
    with open(b_manifest_json_path, 'rb') as manifest_fd:
IOError: [Errno 2] No such file or directory: '/tmp/tmp.pOHC9vpM6d/local/ansible_collections/ansible_test/development/MANIFEST.json'
```

mkrizek commented:
!waffling ci_verified

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -105,3 +108,14 @@ def get_collection_dependencies(self, collection_candidate):
         )
+
+
+def _http_404_to_empty(func):
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except GalaxyError as glxy_err:
+            if glxy_err.http_code == _HTTP_NOT_FOUND:
+                return []
+            raise
+    return wrapper
```
@jborean93 I seem to recall you wanted to move this hack to the API module

webknjaz committed:
Add string representation methods to GalaxyAPI

webknjaz committed:
Make GalaxyAPI instances sortable

webknjaz committed:
Drop asserts from the `collection install` CLI

webknjaz commented on requirements.txt:
```diff
@@ -7,3 +7,4 @@ jinja2
 PyYAML
 cryptography
 packaging
+resolvelib >= 0.5.3  # New dependency resolver that pip>=20.3 uses
```
@s-hertel @jborean93 https://github.com/sarugaku/resolvelib/issues/69 seems to indicate that 0.6 will have breaking changes so we probably need to cap the version too...

webknjaz commented on test/integration/targets/ansible-galaxy-collection-scm/tasks/scm_dependency_deduplication.yml:
```diff
@@ -30,16 +49,14 @@
-      - "'namespace_2.collection_2' in command.stdout_lines[6]"
-      - "'namespace_2.collection_2' in command.stdout_lines[7]"
-      - "'namespace_2.collection_2' in command.stdout_lines[8]"
+      - command.stdout_lines | length == 12
+      - command.stdout_lines[5] == "Starting collection install process"
+      - "'namespace_1.collection_1' in command.stdout_lines[6]"
+      - "'namespace_1.collection_1' in command.stdout_lines[7]"
+      - "'namespace_1.collection_1' in command.stdout_lines[8]"
+      - "'namespace_2.collection_2' in command.stdout_lines[9]"
+      - "'namespace_2.collection_2' in command.stdout_lines[10]"
+      - "'namespace_2.collection_2' in command.stdout_lines[11]"
```
This may potentially be order-dependent aka flaky, like the first test in this file.

webknjaz replied:
I was right. This happened again: https://dev.azure.com/ansible/ansible/_build/results?buildId=4597&view=logs&j=a00789c5-8412-587b-63fc-93dbdf0cd470&t=e049f74a-b955-5412-bd69-ba41533160f3&l=9242

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,455 @@
+from ansible.utils.display import Display
+
+
+try:  # NOTE: py3/py2 compat
+    # FIXME: put somewhere into compat
+    _is_py_id = str.isidentifier  # type: ignore  # py2 mypy can't deal with it
+except AttributeError:  # Python 2
+    # FIXME: port this to AnsibleCollectionRef.is_valid_collection_name
+    from re import match as _match_pattern
+    from tokenize import Name as _VALID_IDENTIFIER_REGEX
+    _valid_identifier_string_regex = ''.join((_VALID_IDENTIFIER_REGEX, r'\Z'))
+
+    def _is_py_id(tested_str):
+        # Ref: https://stackoverflow.com/a/55802320/595220
+        return bool(_match_pattern(_valid_identifier_string_regex, tested_str))
```
need to move this out

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,455 @@
+# Copyright: (c) 2020-2021, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""Dependency structs."""
+# FIXME: add caching all over the place
```
this'd be great

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,455 @@
+
+
+def _is_fqcn(tested_str):
+    # FIXME: port this to AnsibleCollectionRef.is_valid_collection_name
```
TODO ^

webknjaz replied:
https://github.com/ansible/ansible/pull/73279

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,455 @@
+            format(fqcn=to_text(self.fqcn), ver=to_text(self.ver))
+        )
+
+    def _get_separate_ns_n_name(self):  # FIXME: use LRU cache
```
TODO ^

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,455 @@
+
+    @property
+    def is_file(self):
+        # FIXME: Make the checks less dynamic assuming that the type is
```
TODO ^

webknjaz commented:
TODO: add a change fragment

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -730,73 +574,107 @@ def validate_collection_path(collection_path):
+        search_paths,  # type: Iterable[str]
+        apis,  # type: Iterable[GalaxyAPI]
+        ignore_errors,  # type: bool
+        allow_pre_release,  # type: bool  # FIXME: is this arg necessary?
```
FIXME ^

webknjaz committed:
Add a change note

webknjaz commented on test/sanity/code-smell/mypy.py:
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python
```
@mattclay I hope it's okay to have this hardcoded for now.

mattclay commented on requirements.txt:
```diff
@@ -7,3 +7,7 @@ jinja2
+# NOTE: resolvelib 0.x version bumps should be considered major/breaking
+# NOTE: and we should update the upper cap with care, at least until 1.0
+# NOTE: Ref: https://github.com/sarugaku/resolvelib/issues/69
+resolvelib >= 0.5.3, < 0.6.0  # New dependency resolver that pip>=20.3 uses
```
Constraints included here need to be in `test/lib/ansible_test/_data/requirements/constraints.txt` also.

webknjaz replied:
Oh, I missed that. Should we have some sanity check for this?

mattclay replied:
No, because we haven't had any constraints in our requirements file.

We'll probably want to add one that makes sure any constraints here match the constraints file.

mattclay commented on requirements.txt:
```diff
@@ -7,3 +7,7 @@ jinja2
+# NOTE: resolvelib 0.x version bumps should be considered major/breaking
+# NOTE: and we should update the upper cap with care, at least until 1.0
+# NOTE: Ref: https://github.com/sarugaku/resolvelib/issues/69
+resolvelib >= 0.5.3, < 0.6.0  # New dependency resolver that pip>=20.3 uses
```
```suggestion
resolvelib >= 0.5.3, < 0.6.0  # dependency resolver used by ansible-galaxy
```

webknjaz replied:
@mattclay I wanted to mention that it's the same as pip uses

mattclay replied:
That's fine to mention in the PR or docs, but why is it relevant in the requirements file?

webknjaz replied:
fair enough

mattclay reviewed (dismissed):
I'd like to see the addition of mypy as a sanity test done in a separate PR, so we can focus on the galaxy changes first.

I haven't reviewed the non-test changes yet.

webknjaz commented:
@mattclay 

> I'd like to see the addition of mypy as a sanity test done in a separate PR, so we can focus on the galaxy changes first.

No problem. I just needed it here during development. Here you go: https://github.com/ansible/ansible/pull/73319.

I think I've addressed all your requests now.

webknjaz commented on test/lib/ansible_test/_data/requirements/constraints.txt:
```diff
@@ -1,3 +1,4 @@
+resolvelib >= 0.5.3, < 0.6.0  # keep in sync with `requirements.txt`
```
@mattclay I wasn't sure if I was supposed to put it in any specific part of this file. Is this location fine?

felixfontein commented:
If I install this branch in an empty docker container and run
```
ansible-galaxy collection install community.network
ansible-galaxy collection verify community.network
```
the second fails with
```
ERROR! Error when getting collection version metadata for community.network:* from default (https://galaxy.ansible.com/api/) (HTTP Code: 404, Message: Not found. Code: not_found)
```

felixfontein commented:
When running `ansible-galaxy collection list` from this branch on my system (where some collections are checked out repos, not all having a version in galaxy.yml), it crashes:
```
# /path/to/ansible_collections
Collection                              Version  
--------------------------------------- ---------
amazon.aws                              1.3.0    
ERROR! Unexpected Exception, this is probably a bug: unsupported format string passed to NoneType.__format__
the full traceback was:

Traceback (most recent call last):
  File "/path/to/ansible/bin/ansible-galaxy", line 135, in <module>
    exit_code = cli.run()
  File "/path/to/ansible/lib/ansible/cli/galaxy.py", line 542, in run
    context.CLIARGS['func']()
  File "/path/to/ansible/lib/ansible/cli/galaxy.py", line 1308, in execute_list
    self.execute_list_collection()
  File "/path/to/ansible/lib/ansible/cli/galaxy.py", line 74, in method_wrapper
    return wrapped_method(*args, **kwargs)
  File "/path/to/ansible/lib/ansible/cli/galaxy.py", line 1452, in execute_list_collection
    _display_collection(collection, fqcn_width, version_width)
  File "/path/to/ansible/lib/ansible/cli/galaxy.py", line 101, in _display_collection
    display.display('{fqcn:{cwidth}} {version:{vwidth}}'.format(
TypeError: unsupported format string passed to NoneType.__format__
```

webknjaz commented:
@felixfontein Woah! That's unexpected. I guess we need more integration tests. Thanks for reporting! Is that under Python 2?

webknjaz commented:
Fix for the first one would be
```diff
diff --git a/lib/ansible/galaxy/collection/__init__.py b/lib/ansible/galaxy/collection/__init__.py
index 641e88e1d5a..21e09503fdf 100644
--- a/lib/ansible/galaxy/collection/__init__.py
+++ b/lib/ansible/galaxy/collection/__init__.py
@@ -623,7 +623,7 @@ def verify_collections(
                     raise AnsibleError(message='Collection %s is not installed in any of the collection paths.' % collection.fqcn)
 
                 remote_collection = Candidate(
-                    collection.fqcn, collection.ver,
+                    local_collection.fqcn, local_collection.ver,
                     None, 'galaxy',
                 )
```

webknjaz commented:
I tried to reproduce the second bug with `list` but the fallback works (under Python 3.8.5):
```console
$ mv -v ansible-collections-here/ansible_collections/amazon/aws/MANIFEST.json{,.bak}
renamed 'ansible-collections-here/ansible_collections/amazon/aws/MANIFEST.json' -> 'ansible-collections-here/ansible_collections/amazon/aws/MANIFEST.json.bak'
$ mv -v ansible-collections-here/ansible_collections/amazon/aws/FILES.json{,.bak}   
renamed 'ansible-collections-here/ansible_collections/amazon/aws/FILES.json' -> 'ansible-collections-here/ansible_collections/amazon/aws/FILES.json.bak'
$ ansible-galaxy collection list --collections-path ansible-collections-here     

[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying
out features under development. This is a rapidly changing source of code and can become unstable at any point.
[WARNING]: Collection at '~/src/experiments/ansible-galaxy-collection-resolver/collections/ansible-collections-here/ansible_collections/amazon/aws'
does not have a MANIFEST.json file, nor has it galaxy.yml: cannot detect version.

# ~/src/experiments/ansible-galaxy-collection-resolver/collections/ansible-collections-here/ansible_collections
Collection           Version   
-------------------- ----------
amazon.aws           *         
ansible.netcommon    1.4.2-dev1
community.general    1.3.0     
community.kubernetes 1.1.1     
community.mysql      1.1.1     
google.cloud         1.0.1     

# ~/.ansible/collections/ansible_collections
Collection           Version
-------------------- -------
amazon.aws           1.3.0  
ansible.netcommon    1.4.1  
ansible.posix        1.1.1  
community.general    1.3.2  
community.kubernetes 1.1.1
google.cloud         1.0.1
```

I thought it'd explode with no metadata but it was fine...

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -76,20 +99,19 @@ def _display_role(gr):
-        fqcn=to_text(collection),
-        version=collection.latest_version,
+        fqcn=to_text(collection.fqcn),
+        version=collection.ver,
```
@felixfontein could you check if making this
```suggestion
        version=collection.ver or '*',
```
would fix `list` in your env? It'd be great to know how to reproduce/debug this.

felixfontein replied:
Yes, that actually fixes it. The collection printed there is `ansible.netcommon`, which is the latest git checkout from https://github.com/ansible-collections/ansible.netcommon

Maybe it's because it has an explicit `version: null` in its galaxy.yml: https://github.com/ansible-collections/ansible.netcommon/blob/main/galaxy.yml#L13. But when I comment `version: null` out, `ansible-galaxy collection list` simply will not work (both in `devel` and this branch):
```
ERROR! The collection galaxy.yml at '/path/to/ansible_collections/ansible/netcommon/galaxy.yml' is missing the following mandatory keys: version
```

webknjaz replied:
> Maybe it's because it has an explicit `version: null` in its galaxy.yml

Oh, that would make sense. `dict.get('version', '*')` wouldn't fall back to `*` in case if the key exists but is `None`. Thanks for the pointer!

felixfontein replied:
That should better be written as `dict.get('version') or '*'`.

webknjaz replied:
Yeah, probably. I'm just trying to figure out on what level this should be fixed and if there are other cases where this problem would happen if I just patch here.

samdoran replied:
I attempted to fix this in https://github.com/ansible/ansible/pull/70185, but that ended up being way too invasive. Since `_display_collection()` is only used when listing collection information, I think it's fine to do `collection.get(ver) or '*'` here.

webknjaz replied:
@samdoran it's already been fixed elsewhere to make sure that missing or falsy version is turned into `*` when the metadata is read.

felixfontein commented:
@webknjaz it's all under Python 3.8 (docker container) and 3.9 (my system). I'll try to debug the `list` issue.

webknjaz commented:
Looks like the `verify` bugfix is slightly more complicated:
```diff
diff --git a/lib/ansible/galaxy/collection/__init__.py b/lib/ansible/galaxy/collection/__init__.py
index 641e88e1d5a..e28655b7b46 100644
--- a/lib/ansible/galaxy/collection/__init__.py
+++ b/lib/ansible/galaxy/collection/__init__.py
@@ -623,7 +623,9 @@ def verify_collections(
                     raise AnsibleError(message='Collection %s is not installed in any of the collection paths.' % collection.fqcn)
 
                 remote_collection = Candidate(
-                    collection.fqcn, collection.ver,
+                    collection.fqcn,
+                    collection.ver if collection.ver != '*'
+                    else local_collection.ver,
                     None, 'galaxy',
                 )
 
diff --git a/test/integration/targets/ansible-galaxy-collection/tasks/verify.yml b/test/integration/targets/ansible-galaxy-collection/tasks/verify.yml
index 857426d8422..eaef2f8227b 100644
--- a/test/integration/targets/ansible-galaxy-collection/tasks/verify.yml
+++ b/test/integration/targets/ansible-galaxy-collection/tasks/verify.yml
@@ -40,6 +40,11 @@
       - verify is success
       - "'Collection ansible_test.verify contains modified content' not in verify.stdout"
 
+- name: verify the installed collection against the server, with unspecified version in CLI
+  command: ansible-galaxy collection verify ansible_test.verify
+  environment:
+    ANSIBLE_COLLECTIONS_PATH: '{{ galaxy_dir }}'
+
 - name: verify a collection that doesn't appear to be installed
   command: ansible-galaxy collection verify ansible_test.verify:1.0.0
   register: verify
```

webknjaz commented:
@felixfontein FTR the last commit https://github.com/ansible/ansible/commit/d4886ee2bd1d3015efe1335aef668998f365d218 fixes that second bug and will be absorbed into an earlier commit later. Meanwhile, @s-hertel is working on regression tests for this case and we're trying to decide how to validate `version` better. The patch allows falsy `version` values in `galaxy.yml` but not in `MANIFEST.json`. It could be improved by actually parsing it with SemVer but I'm not sure if it's worth it.

felixfontein commented:
@webknjaz with the current version of this branch, both bugs are fixed for me.

s-hertel commented:
Some tests are here https://github.com/webknjaz/ansible/pull/359, https://github.com/webknjaz/ansible/pull/360 which will probably be merged on Monday.

s-hertel commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1146,263 +986,159 @@ def _build_collection_dir(b_collection_path, b_collection_output, collection_man
+    """
+    collection_meta = artifacts_manager.get_direct_collection_meta(collection)
+
+    if 'build_ignore' not in collection_meta:  # installed collection, not src
+        # FIXME: optimize this? use a different process? copy instead of build?
+        collection_meta['build_ignore'] = []
```
On devel installing from source control requires a galaxy.yml because it isn't intended for anything but development. We might want to keep this a hard failure.

On devel the error looks like
```
ERROR! git+file:///path/to/test_src/.git#collection_dir appears to be an SCM collection source, but the required galaxy.yml was not found. Append #path/to/collection/ to your URI (before the comma separated version, if one is specified) to point to a directory containing the galaxy.yml or directories of collections
```

s-hertel replied:
The subdirs checking already only handles `galaxy.yml` metadata, and dirs with a MANIFEST.json are ignored.

webknjaz replied:
@s-hertel can this happen when the requirement is constructed via `from_dir_path_implicit()`?

webknjaz replied:
I seem to recall some corner case that caused me to add this. I think it was something like `galaxy.yml` is present by doesn't have `build_ignore`.

s-hertel replied:
I think that cornercase should be fixed. galaxy.yml files without build_ignore are handled by `_normalize_galaxy_yml_manifest()` so the key should be present. Took a stab at fixing this: https://github.com/webknjaz/ansible/pull/361

s-hertel commented on lib/ansible/galaxy/collection/concrete_artifact_manager.py:
```diff
@@ -0,0 +1,647 @@
+
+
+def _extract_collection_from_git(repo_url, coll_ver, b_path):
+    # FIXME: implement fragment_dir?
```
This looks like it was addressed - is fragment_dir something different than the current fragment handling?

webknjaz replied:
No idea what past me meant by this. You're probably correct.
```suggestion
```

s-hertel commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,433 @@
+                    tip = dir_tip_tmpl.format(src=req_source)
+                elif req_name is not None and os.path.isdir(req_name):
+                    tip = dir_tip_tmpl.format(src=req_name)
+                else:
```
Including a tip for `name` would be helpful for large requirement files with misspelled tar.gz files.
```suggestion
                elif req_name:
                    tip = '\n\nCould not find {0}.'.format(req_name)
                else:
```

s-hertel commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -0,0 +1,274 @@
+        conflict with each other (because they have same name but
+        different versions).
+        """
+        # FIXME: what to do if it is None? Rely on src? Use the whole tuple?
```
I don't see a scenario when this would be None.

webknjaz replied:
Yeah... This used to rely on FQCN which is `None` for virtual/scm cases but since that caused unwanted behavior I changed it. I guess this comment can be dropped now.
```suggestion
```

s-hertel reviewed (approved):
This is awesome :smile: I had a couple minor comments, but overall looks ready to merge.

samdoran commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -0,0 +1,435 @@
+    # FIXME: port this to AnsibleCollectionRef.is_valid_collection_name
+    from re import match as _match_pattern
+    from tokenize import Name as _VALID_IDENTIFIER_REGEX
+    _valid_identifier_string_regex = ''.join((_VALID_IDENTIFIER_REGEX, r'\Z'))
+
+    def _is_py_id(tested_str):
+        # Ref: https://stackoverflow.com/a/55802320/595220
+        return bool(_match_pattern(_valid_identifier_string_regex, tested_str))
```
That's a really interesting piece of information I did not know.

samdoran commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -556,34 +225,49 @@ def build_collection(collection_path, output_path, force):
+        ver=collection_meta['version'],
+    )
+    b_collection_output = os.path.join(
+        output_path,
```
This should be `b_output_path` to indicate it is a series of bytes. Even if the caller is sending in bytes, this function should convert it to bytes just like `b_collection_path` at the beginning. That allows our `b_` naming convention to hold, making this code look incorrect (trying to join `bytes` and `str`).

webknjaz replied:
Addressed.

samdoran commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -594,41 +278,86 @@ def download_collections(collections, output_path, apis, validate_certs, no_deps
-            with open(to_bytes(requirements_path, errors='surrogate_or_strict'), mode='wb') as req_fd:
-                req_fd.write(to_bytes(yaml.safe_dump({'collections': requirements}), errors='surrogate_or_strict'))
+            b_dest_path = os.path.join(
+                to_bytes(output_path, errors='surrogate_or_strict'),
```
Can this use `b_output_path` from line 291?
```suggestion
                b_output_path,
```

samdoran reviewed (approved):
There's _a lot_ in this PR that I don't fully follow. I only saw a few minor things.

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -58,495 +129,93 @@
 
 
 def build_collection(collection_path, output_path, force):
+    # type: (bytes, bytes, bool) -> Text
```
For public APIs we should be accepting text strings and internally convert to bytes where needed. If we really must accept bytes (which in this case I don't see a reason to) we should prefix the arg with `b_` to indicate that.

The `b_` prefix may go away as we start to add type annotations but so far the behaviour is the standard we usually follow in ansible.

webknjaz replied:
Yeah, I guess it's not really public: I think it's not used outside this module.

webknjaz replied:
Addressed.

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -58,495 +129,93 @@
-                pre_release_msg = ''
+    display.vvv("Verifying '{coll!s}'.".format(coll=local_collection))
+    display.vvv(
+        "Installed collection found at '{path!s}'".
+        format(path=to_native(local_collection.src)),
```
Display should use text strings, will need a to_text in the format and u prefix the string so we don't have any ASCII encoding errors on Python 2.

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -58,495 +129,93 @@
+        format(path=to_native(local_collection.src)),
+    )
+    display.vvv(
+        "Remote collection cached as '{path!s}'".
+        format(path=to_native(b_temp_tar_path)),
```
to_text and u prefix on string.

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -594,41 +278,86 @@ def download_collections(collections, output_path, apis, validate_certs, no_deps
-                    b_temp_download_path = requirement.download(b_temp_path)
-                    shutil.move(b_temp_download_path, to_bytes(dest_path, errors='surrogate_or_strict'))
+            display.display(
+                "Downloading collection '{coll!s}' to '{path!s}'".
+                format(coll=concrete_coll_pin, path=to_native(b_output_path)),
```
to_text and prefix the string with u.

webknjaz replied:
```suggestion
                u"Downloading collection '{coll!s}' to '{path!s}'".
                format(coll=to_text(concrete_coll_pin), path=to_text(b_output_path)),
```

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -594,41 +278,86 @@ def download_collections(collections, output_path, apis, validate_certs, no_deps
+            requirements_path, errors='surrogate_or_strict',
+        )
+        display.display(
+            'Writing requirements.yml file of downloaded collections '
+            "to '{path!s}'".format(path=requirements_path),
```
probably want a to_text and u prefix.

webknjaz replied:
```suggestion
            u'Writing requirements.yml file of downloaded collections '
            "to '{path!s}'".format(path=to_text(requirements_path)),
```

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1097,12 +930,19 @@ def reset_stat(tarinfo):
                 # Dealing with a normal file, just add it by name.
-                tar_file.add(os.path.realpath(b_src_path), arcname=filename, recursive=False, filter=reset_stat)
+                tar_file.add(
+                    to_native(os.path.realpath(b_src_path)),
```
is there a reason why you have a native str, I would have assumed it was a byte str beforehand for a reason. I know the tarfile is very particular about the type of string being used.

webknjaz replied:
I don't remember why this was necessary. I seem to recall it's unicode under py3...

webknjaz replied:
Oh, wait! I think I remember. It was mypy that hinted that the type is not right and so I checked this in docs / stubs / source / https://github.com/python/typeshed/blob/869238e/stdlib/2and3/tarfile.pyi#L162-L189 / https://github.com/python/cpython/blob/master/Lib/tarfile.py and this is what was necessary. (str under py2 and StrPath that resolves to Text or PathLike: https://github.com/python/typeshed/blob/16ae4c61201cd8b96b8b22cdfb2ab9e89ba5bcf2/stdlib/_typeshed/__init__.pyi#L59-L63)

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1097,12 +930,19 @@ def reset_stat(tarinfo):
+                )
 
-        shutil.copy(b_tar_filepath, b_tar_path)
+        shutil.copy(to_native(b_tar_filepath), to_native(b_tar_path))
```
Same here, don't we usually use byte strings for file operations.

webknjaz replied:
Same, it accepts `StrPath`: https://github.com/python/typeshed/blob/869238e587ce2bb2e1b990026d5a038f6eb0f1a3/stdlib/2and3/shutil.pyi#L55-L62

jborean93 commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1097,12 +930,19 @@ def reset_stat(tarinfo):
         collection_name = "%s.%s" % (collection_manifest['collection_info']['namespace'],
                                      collection_manifest['collection_info']['name'])
+        tar_path = to_text(b_tar_path)
         display.display('Created collection for %s at %s' % (collection_name, to_text(b_tar_path)))
```
may as well change this to tar_path now that you have converted it above

```suggestion
        display.display(u'Created collection for %s at %s' % (collection_name, tar_path))
```

jborean93 commented on requirements.txt:
```diff
@@ -7,3 +7,7 @@ jinja2
+# NOTE: resolvelib 0.x version bumps should be considered major/breaking
+# NOTE: and we should update the upper cap with care, at least until 1.0
+# NOTE: Ref: https://github.com/sarugaku/resolvelib/issues/69
+resolvelib >= 0.5.3, < 0.6.0  # dependency resolver used by ansible-galaxy
```
So do we run the risk here that pip has it's own pinned resolvelib version but we have another or does pip vendor it entirely and we don't care about it?

webknjaz replied:
It's one of the libs pip re-vendors every once in a while: https://github.com/pypa/pip/tree/master/src/pip/_vendor#vendoring-policy. For example: https://github.com/pypa/pip/pull/9369.

webknjaz replied:
So I'd say we don't care. We'll need to upgrade the upper bound once in a while along with updating our code to work with newer public API when necessary.

abadger replied:
Not entirely correct.... Linux distros typically unvendor libraries.  So you could still end up with a problem if you pin your versions too strictly.

webknjaz replied:
@abadger in this case we know for sure that v0.6.0 will add extra arguments w/o defaults to some hooks 100% breaking the API so before changing the upper bound, we'll have to adjust that. Besides, we now have a designated location in the tree where the packagers can drop-in libs w/o having to package them separately if they wish so.

But I see your point: deb packages infamous for regularly breaking pip and other Python packaging-related things. Ref: https://gist.github.com/tiran/2dec9e03c6f901814f6d1e8dad09528e.

Anyway... We have to specify what exactly is supported right now, this info gets into the metadata and the packagers are responsible for not breaking stuff. If we say that 0.6.0 is not supported and they ignore the instruction, it's not our fault, is it?

jborean93 reviewed (commented):
Some encoding questions but mostly superficial. Looks great otherwise.

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -58,495 +129,93 @@
-            else:
-                pre_release_msg = ''
+    display.vvv("Verifying '{coll!s}'.".format(coll=local_collection))
+    display.vvv(
+        "Installed collection found at '{path!s}'".
+        format(path=to_native(local_collection.src)),
+    )
+    display.vvv(
+        "Remote collection cached as '{path!s}'".
+        format(path=to_native(b_temp_tar_path)),
+    )
```
```suggestion
    display.vvv(
        u"Installed collection found at '{path!s}'".
        format(path=to_text(local_collection.src)),
    )
    display.vvv(
        u"Remote collection cached as '{path!s}'".
        format(path=to_text(b_temp_tar_path)),
    )
```

abadger commented on lib/ansible/galaxy/api.py:
```diff
@@ -233,11 +234,17 @@ def __init__(self, namespace, name, version, download_url, artifact_sha256, depe
         self.dependencies = dependencies
 
 
+@functools.total_ordering
```
total_ordering isn't present in 2.7 so you can't use this until ansible-2.12.

sivel replied:
Not technically.  We do not support 2.6 on the controller, and as long as it isn't causing testing issues, I see no reason why not.  2.7 is our minimum required for the controller.

abadger replied:
Ah, sorry.  I checked the python 3 docs and saw it wasn't present from the outset.  I forgot that python2.7's release came after python 3.1 so total_ordering might have been added to both releases.
